src/Agent.Worker/CommandHandler.cs

```diff
@@ -0,0 +1,63 @@
+﻿using Microsoft.TeamFoundation.DistributedTask.WebApi;
+using Microsoft.VisualStudio.Services.Agent.Util;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker
+{
+    [ServiceLocator(Default = typeof(CommandHandler))]
+    public interface ICommandHandler : IAgentService
+    {
+        bool IsCommand(string input, out Command command);
+
+        void ProcessCommand(IExecutionContext context, Command command);
+    }
+
+    public class CommandHandler : AgentService, ICommandHandler
+    {
+        private readonly Dictionary<String, ICommandExtension> _commandHandlers = new Dictionary<String, ICommandExtension>(StringComparer.OrdinalIgnoreCase);
+
+        public override void Initialize(IHostContext hostContext)
+        {
+            base.Initialize(hostContext);
+
+            // Register all command extensions
+            var extensionManager = hostContext.GetService<IExtensionManager>();
+            foreach (var commandExt in extensionManager.GetExtensions<ICommandExtension>() ?? new List<ICommandExtension>())
+            {
+                Trace.Info($"Register command extension for area {commandExt.CommandArea}");
+                _commandHandlers[commandExt.CommandArea] = commandExt;
+            }
+        }
+
+        public bool IsCommand(string message, out Command command)
+        {
+            return Command.TryParse(message, out command);
```

Comment:
What if someone has a malformed command that won't parse but they tried?  With this, it would just  ignore.  What if IsCommand returned startsWith ##vso[ ??  Then the first thing Process does is a TryParse - if that fails, you can add a warning issue about invalid command.  That signals they have an issue earlier instead of them just having to notice that the work they thought they invoked didn't happen.